library/alloc/src/collections/btree/node.rs

```diff
@@ -219,6 +223,9 @@ impl<K, V> NodeRef<marker::Owned, K, V, marker::LeafOrInternal> {
 ///      although insert methods allow a mutable pointer to a value to coexist.
 ///    - When this is `Owned`, the `NodeRef` acts roughly like `Box<Node>`,
 ///      but does not have a destructor, and must be cleaned up manually.
+///    - When this is `Dying`, the `NodeRef` still acts roughly like `Box<Node>`,
+///      but has methods to destroy the tree bit by bit, no methods to extend or
+///      change the tree, and traversal methods implicitly become unsafe.
```

Comment:
We should be clearer about this implicit unsafety - are the methods actually unsafe to call (in the "unsafe fn" sense)? If not, then we should say something like "traversal methods are not annotated as unsafe, but in practice are unsafe" or so.

Reply:
They're not actually unsafe, and it would be annoying if something like `left_kv` or `first_leaf_edge` was marked unsafe just because the borrow type might be Owned/Dying. More so for functions we never ever use them for that borrow type. We could have wrappers around a private, generic `BorrowType` implementation though, and then I'm pretty sure the Owned/Dying wrapper could be marked unsafe. Not sure how many more functions that gives and whether it helps understanding.

Reply:
I don't know what you mean by this then. The added comment indicates that these methods are somehow unsafe, but then you say they are not. I'm not asking for them to be marked unsafe here, I'm trying to make this comment be clearer about what "implicitly become unsafe" means.

Reply:
I thought with "actually unsafe" you meant declared as `unsafe fn`. `first_leaf_edge` is not marked unsafe, but it "is" definitely unsafe: if you were to invoke it on an owned node that you got to by forward iterating enough times and then ascending, you'll descend into a deallocated `InternalNode` , read its former edges, descend into those garbage pointers…

Whether there's any performance advantage in the source of this unsafety, the tactic of deallocating-while-iterating (whenever I pass a bridge, let me already burn the village I leave behind), is another matter.

Reply:
Ok. It would be good to write this comment in a way that indicates this more clearly, for example:

> ... and traversal methods, while not marked as unsafe to call, can invoke UB if called incorrectly.

The performance advantage (or lack thereof) is something to consider separately, I think.

Reply:
> something to consider separately

Definitely, but this first commit and the safety of dying trees too. I'll keep this PR for the "when Owned means root" theme.

Reply:
I am prepared to r+ this with this comment amended - not sure what else you are planning to do here.

Reply:
Oh, but if you like the idea of those borrow type traits, then it's appealing to define a trait for all reasonable borrow types, excluding Dying, and define `first_leaf_edge` with that trait, so that we're sure it's always genuinely safe. Perhaps ending up with no functions at all that become unsafe.

But the way I defined Dying here, it's already used while descending towards the leaf nodes, while there's nothing unsafe about navigation at that stage. Got to think it through a bit more.

Reply:
I'm not sure I *like* them. I would prefer to avoid adding more traits/markers, etc. if we can, because I'm not sure avoiding the little bit of unsafety actually improves the code as we're also increasing complexity. I'm not unwilling to add them, though, hard to really give a definitive answer.

Reply:
I wrote:
> Perhaps ending up with no functions at all that become unsafe.

That's baloney. `descend` is required during the destruction of the tree (not just to start off at the leaf nodes) and is inherently unsafe when part of the edge arrays contains stale pointer and `descend` cannot know what part that is. The only way to avoid that is to define a separate unsafe `descend` (wrapper) function for the Dying type, but then you cannot access that in the generic navigation functions being used. I think the only way is to write the whole destructive iterator code with parallel unsafe functions. It doesn't look like hugely more code than the part that already is parallel (the part defined by `def_next_kv_uncheched_dealloc`), but not worth the trouble.

Reply:
I am strongly against duplicating code for one unsafe and one safe version.